src/components/implementation/tests/micro_booter/mb_perf_tests.c

```diff
@@ -0,0 +1,237 @@
+            perfdata_99ptile(&pd[cos_cpuid()]));
+
+    perfdata_init(&pd[cos_cpuid()], "Async Endpoints => One Way");
+
+    for (i = 0; i < ITER; i++) {
+        rdtscll(s);
+        cos_asnd(sc, 1);
+        rdtscll(e);
+
+        perfdata_add(&pd[cos_cpuid()], (e - s));
+    }
+
+    perfdata_calc(&pd[cos_cpuid()]);
+
+    PRINTC("\tAsync Endpoints => One Way:\t\tAVG:%llu, MAX:%llu, MIN:%llu, ITER:%d\n",
+            perfdata_avg(&pd[cos_cpuid()]), perfdata_max(&pd[cos_cpuid()]), perfdata_min(&pd[cos_cpuid()]),
+            perfdata_sz(&pd[cos_cpuid()]));      
+
+    printc("\t\t\t\t\t\t\tSD:%llu, 90%%:%llu, 95%%:%llu, 99%%:%llu\n",
+            perfdata_sd(&pd[cos_cpuid()]),perfdata_90ptile(&pd[cos_cpuid()]), perfdata_95ptile(&pd[cos_cpuid()]),
+            perfdata_99ptile(&pd[cos_cpuid()]));
+
+    async_test_flag_[cos_cpuid()] = 0;
+    while (1) cos_thd_switch(tc);
+}
+
+static void
+test_async_endpoints_perf(void)
+{
+    thdcap_t  tcp, tcc;
+    tcap_t    tccp, tccc;
+    arcvcap_t rcp, rcc;
+
+    /* parent rcv capabilities */
+    tcp = cos_thd_alloc(&booter_info, booter_info.comp_cap, async_thd_parent_perf,
+                        (void *)BOOT_CAPTBL_SELF_INITTHD_CPU_BASE);
+    if(EXPECT_LL_LT(1, tcp, "Test Async Endpoints")) return;
+    tccp = cos_tcap_alloc(&booter_info);
+    if(EXPECT_LL_LT(1, tccp, "Test Async Endpoints")) return;
+    rcp = cos_arcv_alloc(&booter_info, tcp, tccp, booter_info.comp_cap, BOOT_CAPTBL_SELF_INITRCV_CPU_BASE);
+    if(EXPECT_LL_LT(1, rcp, "Test Async Endpoints")) return;
+    if(EXPECT_LL_NEQ(0,cos_tcap_transfer(rcp, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, TCAP_RES_INF,
+        TCAP_PRIO_MAX + 1), "Test Async Endpoints")) {
+        return;
+    }
+
+    /* child rcv capabilities */
+    tcc = cos_thd_alloc(&booter_info, booter_info.comp_cap, async_thd_fn_perf, (void *)tcp);
+    if(EXPECT_LL_LT(1, tcc, "Test Async Endpoints")) return;
+    tccc = cos_tcap_alloc(&booter_info);
+    if(EXPECT_LL_LT(1, tccc, "Test Async Endpoints")) return;
+    rcc = cos_arcv_alloc(&booter_info, tcc, tccc, booter_info.comp_cap, rcp);
+    if(EXPECT_LL_LT(1, rcc, "Test Async Endpoints")) return;
+    if(EXPECT_LL_NEQ(0,cos_tcap_transfer(rcc, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, TCAP_RES_INF,
+        TCAP_PRIO_MAX), "Test Async Endpoints"))
+         return;
+
+    /* make the snd channel to the child */
+    scp_global[cos_cpuid()] = cos_asnd_alloc(&booter_info, rcc, booter_info.captbl_cap);
+    if(EXPECT_LL_EQ(0, scp_global[cos_cpuid()], "Test Async Endpoints")) return;
+
+    /* make the snd channel to the parent */
+    scc_global[cos_cpuid()] = cos_asnd_alloc(&booter_info, rcp, booter_info.captbl_cap);
+    if(EXPECT_LL_EQ(0, scp_global[cos_cpuid()], "Test Async Endpoints")) return;
+
+    rcc_global[cos_cpuid()] = rcc;
+    rcp_global[cos_cpuid()] = rcp;
+
+    async_test_flag_[cos_cpuid()] = 1;
+    while (async_test_flag_[cos_cpuid()]) cos_thd_switch(tcp);
+}
+
+void
+test_print_ubench(void)
```

Comment:
The issue is the way the measurements are taken after the unit test is successful. Since we want to delay the prints this is the way to store each individual measurement.

Reply:
You're making a false dichotomy, and ignoring the other option: pull the performance results out, and save them in a smaller struct until print out, and reuse the backing, large data-structure. Don't assume that the only way to solve the problem is to avoid doing the delayed printout.

Reply:
I had a smaller struct to save the data but you told me to remove from the existence of the perfdata.h